### PR TITLE
ROX-9813 set enforcement not supported for Network Policy fields

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

Disallow enforcement of Network Policy System Policy fields.

Right now we don't support enforcement of these kind of fields. If enforcement is set we return an error at the creation time.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* [x] Unit testing
* [x] Manual testing

Creating a System Policy like:

```json
{
	"name": "Deployments should have ingress and egress Network Policy",
	"severity": "MEDIUM_SEVERITY",
	"disabled": false,
	"enforcementActions": [
	    "SCALE_TO_ZERO_ENFORCEMENT"
	    ],
	"lifecycleStages": [
		"DEPLOY"
	],
	"eventSource": "NOT_APPLICABLE",
	"isDefault": false,
	"categories": [
		"Anomalous Activity"
	],
	"policySections": [
		{
			"sectionName": "Policy Section 1",
			"policyGroups": [
				{
					"fieldName": "Missing Ingress Network Policy",
					"values": [
						{
							"value": "true"
						}
					]
				},
				{
					"fieldName": "Missing Egress Network Policy",
					"values": [
						{
							"value": "true"
						}
					]
				}
			]
		}
	]
}
```

Should return an error:

```json
{
  "error": "policy invalid error: enforcement of Missing Ingress Network Policy is not allowed: invalid arguments",
  "code": 3,
  "message": "policy invalid error: enforcement of Missing Ingress Network Policy is not allowed: invalid arguments",
  "details": []
}
```

### CODE REVIEW

**Review by vikin91 — APPROVED** [review]
Looks good!

Are there any tests that could be added here?

**Comment by lvalerom** [comment]
> Are there any tests that could be added here?

I didn't add any because this is a provisional change and AFAIK we will allow enforcement in the future. But if you think it would be better to add them I can do that.

**Comment by vikin91** [comment]
Well, would be cool to have some if they are not so difficult to add here. I think we will need them for enforcement definitely, but not sure about now. In my opinion this could be merged the way it is. Let's see what others think

File: central/policy/service/validator.go [hunk 1]
```diff
@@ -87,6 +89,9 @@ func (s *policyValidator) internalValidate(policy *storage.Policy, additionalVal
 	errorList.AddError(s.validateExclusions(policy))
 	errorList.AddError(s.validateCapabilities(policy))
 	errorList.AddError(s.validateEventSource(policy))
+	if features.NetworkPolicySystemPolicy.Enabled() {
+		errorList.AddError(s.validateNetworkPolicyEnforcement(policy))
```
Comment (fredrb) [inline central/policy/service/validator.go:93]: Interesting that we don't have other policies that aren't enforced.

This is more of a matter of personal preference, but I'd be more inclined to do this in a generic way to make it easier to add not enforceable policies in the future. I was hoping that making a policy not enforceable would be as simple as adding a new field name to a map, for example. 

We could achieve this by creating such a map of `notEnforceablePolicies`, and use as:

```go
// (without feature flag check)
errorList.AddError(s.validateEnforcement(policy))
```

```go
func (s *policyValidator) validateEnforcement(policy *storage.Policy) error {
        if len(policy.GetEnforcementActions()) > 0 {
		for _, s := range policy.GetPolicySections() {
			for _, g := range s.GetPolicyGroups() {
				if _, ok := s.nonEnforceablePolicies[g.GetFieldName()]; ok {
					return errors.Errorf("enforcement of %s field is not allowed", g.GetFieldName())
				}
			}
		}
	}
	return nil
}
```

**Review by fredrb — APPROVED** [review]

**Comment by fredrb** [comment]
> I didn't add any because this is a provisional change and AFAIK we will allow enforcement in the future.

@lvalerom I wouldn't bet on that, honestly. I think it's better to treat having not enforceable policies as a feature for now, if we happen to disable it in the future: great. But it might as well stay in the application for a long time.

File: central/policy/service/validator.go [hunk 2]
```diff
@@ -377,3 +382,16 @@ func removeEnforcementForLifecycle(policy *storage.Policy, stage storage.Lifecyc
 func (s *policyValidator) isAuditEventPolicy(policy *storage.Policy) bool {
 	return policy.GetEventSource() == storage.EventSource_AUDIT_LOG_EVENT
 }
+
+func (s *policyValidator) validateNetworkPolicyEnforcement(policy *storage.Policy) error {
+	if len(policy.GetEnforcementActions()) > 0 {
+		for _, s := range policy.GetPolicySections() {
+			for _, g := range s.GetPolicyGroups() {
+				if g.GetFieldName() == augmentedobjs.MissingEgressPolicyCustomTag || g.GetFieldName() == augmentedobjs.MissingIngressPolicyCustomTag {
+					return errors.New("enforcement of Network Policies is not allowed yet")
```
Comment (fredrb) [inline central/policy/service/validator.go:391]:
```suggestion
					return errors.New("enforcement of Network Policies is not allowed")
```

**Review by fredrb — COMMENTED** [review]
Given the conversation above

**Comment by lvalerom** [comment]
> > I didn't add any because this is a provisional change and AFAIK we will allow enforcement in the future.
> 
> @lvalerom I wouldn't bet on that, honestly. I think it's better to treat having not enforceable policies as a feature for now, if we happen to disable it in the future: great. But it might as well stay in the application for a long time.

Fair enough. I'll add some tests. I'll also refactor to make the function generic. Thanks @vikin91 and @fredrb for the input.

File: central/policy/service/validator.go (hunk 1 above)
Comment (lvalerom) [reply central/policy/service/validator.go]: Done. For now I'm hardcoding the values of the `nonEnforceablePolicyFields` map in the `policyValidator`. I'm going to take another a look and see there's a way to somehow *mark* a policy field as not enforceable but I haven't seen it so far.

File: central/policy/service/validator.go (hunk 2 above)
Comment (lvalerom) [reply central/policy/service/validator.go]: Changed to the generic function.

**Review by fredrb — APPROVED** [review]

File: central/policy/service/validator.go (hunk 1 above)
Comment (fredrb) [reply central/policy/service/validator.go]: I think it's fine the way you did. There is no policy field structure, there are a lot of different structures linked to each other for each part (field name, metadata, violation message, and so on...). The field name stored in the actual policy is just a generic structure with a name and a list of values.

**Pull request merged**